Use HBAT 200 Data and practice performing multiple linear regression involving TrainTestSPlit and Model Evaluation.

This of the whole process as puzzle pieces that you have to put together in their proper places.

ID

X1 - Customer Type

X2 - Industry Type

X3 - Firm Size

X4 - Region

X5 - Distribution System

X6 - Product Quality

X7 - E-Commerce

X8 - Technical Support

X9 - Complaint Resolution

X10 - Advertising

X11 - Product Line

X12 - Salesforce Image


X13 - Competitive Pricing
X14 - Warranty & Claims

X15 - New Products

X16 - Order & Billing

X17 - Price Flexibility

X18 - Delivery Speed

X19 - Satisfaction

X20 - Likely to Recommend

X21 - Likely to Purchase

X22 - Purchase Level

X23 - Consider Strategic Alliance

Independent Variable / Predictors are quantitative/continuous. The variable can take a on a range of values and it will have meaningful order (eg. age, income, time).

Dependent Variable / Outcome Variables should be quantitative/continuous (eg. sales, height, test scores). Here possibly X19 - Satisfaction, X20 - Likely to Recommend,
X21 - Likely to Purchase, X22 - Purchase Level.

**1. Linear Regression**

Import libraries and read/display .csv

In [ ]:
import pandas as pd
import numpy as np

# Create a DataFrame
df = pd.read_csv('HBAT-200-Data.csv')

# Display the first few rows of the dataset
df.head()

ID  X1  X2  X3  X4  X5   X6   X7   X8   X9  ...  X14  X15  X16  X17  X18  \
0   1   2   0   1   1   1  8.5  3.9  2.5  5.9  ...  4.7  4.3  5.0  5.1  3.7   
1   2   3   1   0   0   0  8.2  2.7  5.1  7.2  ...  5.5  4.0  3.9  4.3  4.9   
2   3   3   0   1   1   1  9.2  3.4  5.6  5.6  ...  6.2  4.6  5.4  4.0  4.5   
3   4   1   1   1   1   0  6.4  3.3  7.0  3.7  ...  7.0  3.6  4.3  4.1  3.0   
4   5   2   0   1   0   1  9.0  3.4  5.2  4.6  ...  6.1  4.5  4.5  3.5  3.5   

   X19  X20  X21   X22  X23  
0  8.2  8.0  8.4  65.1    1  
1  5.7  6.5  7.5  67.1    0  
2  8.9  8.4  9.0  72.1    1  
3  4.8  6.0  7.2  40.1    0  
4  7.1  6.6  9.0  57.1    0  

[5 rows x 24 columns]

Identify target

In [ ]:
# Calculate correlations between all columns (this computes the Pearson correlation coefficients between all pairs of numeric columns in the DataFrame)
correlations = df.corr()

# Identify potential features based on correlation with other columns
# Exclude highly correlated columns and non-numeric columns like 'ID'
exclude_columns = ['ID']

# Filter out non-numeric columns and excluded columns
numeric_columns = df.select_dtypes(include=[np.number]).columns # selects only numeric columns
filtered_features = [col for col in numeric_columns if col not in exclude_columns]

# Determine the target variable
# Find the column with the highest average correlation with other columns
avg_correlations = correlations.abs().mean() # computes mean of absolute correlation values for each column
target_variable = avg_correlations.idxmax() # identifies column w/highest average correlation (idxmax - index max)

# Define the feature matrix X and target variable y
X = df[filtered_features] # creates matrix x w/identified feature colulmns
y = df[target_variable] # sets target y to identified target column

# Display the identified features and target variable
print("\nIdentified Features:")
print(filtered_features)
print("\nTarget Variable:")
print(target_variable)



Identified Features:
['X1', 'X2', 'X3', 'X4', 'X5', 'X6', 'X7', 'X8', 'X9', 'X10', 'X11', 'X12', 'X13', 'X14', 'X15', 'X16', 'X17', 'X18', 'X19', 'X20', 'X21', 'X22', 'X23']

Target Variable:
X19


Target (y) is X19.

What factors influence X19 Satisfaction?

Possibly:<br>X6-Product Quality<br>
X7-E-Commerce<br>
X8-Technical Support<br>X9-Complaint Resolution<br>X10-Advertising<br>X11-Product Line<br>X12-Salesforce Image<br> X13-Competitive Pricing<br>X14-Warranty & Claims<br>X15-New Products<br>X16-Order & Billing<br>X17-Price Flexibility<br>X18-Delivery Speed

##Initial Model:

In [ ]:
import statsmodels.api as sm

# Prepare data for regression
X = df[['X6', 'X7', 'X8', 'X9', 'X10', 'X11', 'X12', 'X13', 'X14', 'X15', 'X16', 'X17', 'X18']]  # IVs
y = df['X19']  # DV

#This time we investigate ratings starting from X6 to X18

# Add a constant term for the intercept
X = sm.add_constant(X)

# Fit the regression model
model = sm.OLS(y, X).fit()

# Print model summary
print(model.summary())

                            OLS Regression Results                            
Dep. Variable:                    X19   R-squared:                       0.804
Model:                            OLS   Adj. R-squared:                  0.790
Method:                 Least Squares   F-statistic:                     58.76
Date:                Thu, 04 Jul 2024   Prob (F-statistic):           1.20e-58
Time:                        03:45:54   Log-Likelihood:                -163.44
No. Observations:                 200   AIC:                             354.9
Df Residuals:                     186   BIC:                             401.1
Df Model:                          13                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         -2.2549      0.789     -2.859      0.0

##Model Selection/Optimization through backward elimination

In [ ]:
import pandas as pd
from sklearn.feature_selection import SequentialFeatureSelector
from sklearn.linear_model import LinearRegression

# Create a linear regression model
model = LinearRegression()

# Create a SequentialFeatureSelector for backward elimination
selector = SequentialFeatureSelector(model, direction='backward', n_features_to_select=None, scoring='neg_mean_squared_error')

# Fit the selector on the entire dataset
selector.fit(X, y)

# Get the selected features
selected_features = X.columns[selector.get_support()]

# Print the selected features
print("Selected Features:", selected_features)

# Fit the final model using the selected features
final_model = LinearRegression().fit(X[selected_features], y)

# Print the summary of the final model
print("Final Model Summary:")
print("R-squared:", final_model.score(X[selected_features], y))

/usr/local/lib/python3.10/dist-packages/sklearn/feature_selection/_sequential.py:211: FutureWarning: Leaving `n_features_to_select` to None is deprecated in 1.0 and will become 'auto' in 1.3. To keep the same behaviour as with None (i.e. select half of the features) and avoid this warning, you should manually set `n_features_to_select='auto'` and set tol=None when creating an instance.
  warnings.warn(


Selected Features: Index(['X6', 'X7', 'X9', 'X11', 'X12', 'X13', 'X17'], dtype='object')
Final Model Summary:
R-squared: 0.8002523036662732


##Model with Selected Features

['X6', 'X7', 'X9', 'X11', 'X12', 'X13', 'X17']

In [ ]:
# Prepare data for regression
X = df[['X6', 'X7', 'X9', 'X11', 'X12', 'X13', 'X17']]  # IVs
y = df['X19']  # DV

# Add a constant term for the intercept
X = sm.add_constant(X)

# Fit the regression model
model = sm.OLS(y, X).fit()

# Print model summary
print(model.summary())

                            OLS Regression Results                            
Dep. Variable:                    X19   R-squared:                       0.800
Model:                            OLS   Adj. R-squared:                  0.793
Method:                 Least Squares   F-statistic:                     109.9
Date:                Thu, 04 Jul 2024   Prob (F-statistic):           1.15e-63
Time:                        03:51:06   Log-Likelihood:                -165.42
No. Observations:                 200   AIC:                             346.8
Df Residuals:                     192   BIC:                             373.2
Df Model:                           7                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         -2.0262      0.512     -3.954      0.0

X9 - Complaint Resolution has P>|t| of 0.100 which is relatively high (> alpha of 0.5) thus we do not consider it statistically significant (not enough evidence to reject null hypothesis). Thus we will remove it.

##Revised Model

['X6', 'X7', 'X11', 'X12', 'X13', 'X17']

In [ ]:
# Prepare data for regression
X = df[['X6', 'X7', 'X11', 'X12', 'X13', 'X17']]  # IVs
y = df['X19']  # DV

# Add a constant term for the intercept
X = sm.add_constant(X)

# Fit the regression model
model = sm.OLS(y, X).fit()

# Print model summary
print(model.summary())

                            OLS Regression Results                            
Dep. Variable:                    X19   R-squared:                       0.797
Model:                            OLS   Adj. R-squared:                  0.791
Method:                 Least Squares   F-statistic:                     126.6
Date:                Thu, 04 Jul 2024   Prob (F-statistic):           3.77e-64
Time:                        03:57:04   Log-Likelihood:                -166.84
No. Observations:                 200   AIC:                             347.7
Df Residuals:                     193   BIC:                             370.8
Df Model:                           6                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         -2.2767      0.492     -4.630      0.0

In [ ]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.feature_selection import SelectKBest, f_regression
from sklearn.metrics import mean_absolute_error, mean_absolute_percentage_error

# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.4, random_state=23)

Train model

In [ ]:
# Train a linear regression model using the selected features
model = LinearRegression()
model.fit(X_train, y_train)

LinearRegression()

Predict (from X_train)

In [ ]:
# Make predictions on the test set
y_pred = model.predict(X_test)

y_pred

array([7.22462434, 6.05456123, 6.93957304, 7.55637317, 7.4317935 ,
       5.7810438 , 5.51310701, 8.78573448, 8.27579141, 6.19213152,
       7.30895445, 7.85526387, 7.05792771, 7.20981249, 6.73263444,
       8.45297258, 6.28078209, 6.07816216, 7.67968752, 6.66331803,
       9.03859826, 8.27832891, 7.31705423, 6.61481766, 4.65092495,
       8.15559854, 7.05792771, 6.19213152, 6.89392877, 6.0925689 ,
       6.68088585, 6.66331803, 4.86248985, 7.68707311, 7.12134792,
       7.59292165, 6.93957304, 6.81110549, 5.00220856, 6.96464527,
       7.37683068, 7.20981249, 9.03859826, 6.68088585, 9.66773874,
       6.30457523, 7.08717024, 6.27301925, 6.90796374, 8.47549205,
       7.62171111, 7.62171111, 5.337814  , 6.81130864, 9.03859826,
       7.08843437, 7.52882842, 4.70083506, 6.27199311, 7.5292261 ,
       4.85601874, 5.18424709, 7.13682886, 6.59102452, 7.74546772,
       7.06308495, 6.0501209 , 6.61481766, 6.81130864, 7.12134792,
       7.70388928, 7.59292165, 5.13121311, 6.05456123, 6.90796

Check r-squared on the train set

In [ ]:
from sklearn.metrics import r2_score
r2_test = r2_score(y_test, y_pred)
print(f"R-squared Score on Test Set: {r2_test}")

y_train_pred = model.predict(X_train)
r2_train = r2_score(y_train, y_train_pred)
print(f"R-squared Score on Training Set: {r2_train}")

R-squared Score on Test Set: 0.6803164326110448
R-squared Score on Training Set: 0.8491576196868241


R-squared Score on Training Set (0.849) ~ 84.9% of variance in target X19 can be explained by the independent variables/features in the model.

R-squared Score on Test Set (0.680) ~ 68% of variance in the target variable and should provide reasonably good fit to test data.

Testing

In [ ]:
y_pred = model.predict(X_test)
y_pred

array([7.22462434, 6.05456123, 6.93957304, 7.55637317, 7.4317935 ,
       5.7810438 , 5.51310701, 8.78573448, 8.27579141, 6.19213152,
       7.30895445, 7.85526387, 7.05792771, 7.20981249, 6.73263444,
       8.45297258, 6.28078209, 6.07816216, 7.67968752, 6.66331803,
       9.03859826, 8.27832891, 7.31705423, 6.61481766, 4.65092495,
       8.15559854, 7.05792771, 6.19213152, 6.89392877, 6.0925689 ,
       6.68088585, 6.66331803, 4.86248985, 7.68707311, 7.12134792,
       7.59292165, 6.93957304, 6.81110549, 5.00220856, 6.96464527,
       7.37683068, 7.20981249, 9.03859826, 6.68088585, 9.66773874,
       6.30457523, 7.08717024, 6.27301925, 6.90796374, 8.47549205,
       7.62171111, 7.62171111, 5.337814  , 6.81130864, 9.03859826,
       7.08843437, 7.52882842, 4.70083506, 6.27199311, 7.5292261 ,
       4.85601874, 5.18424709, 7.13682886, 6.59102452, 7.74546772,
       7.06308495, 6.0501209 , 6.61481766, 6.81130864, 7.12134792,
       7.70388928, 7.59292165, 5.13121311, 6.05456123, 6.90796

Evaluate performance using MAPE, MAE, MSE, RMSE

In [ ]:
from sklearn.metrics import mean_absolute_error, mean_absolute_percentage_error, mean_squared_error

# MAPE
mape = (mean_absolute_percentage_error(y_test, y_pred) * 100)
print("MAPE:", mape)

#MAE
mae = mean_absolute_error(y_test, y_pred)
print("MAE:", mae)

#MSE (Mean Squared Error)
mse = mean_squared_error(y_test, y_pred)
print("MSE:", mse)

#RMSE (Root Mean squared error)
rmse = np.sqrt(mse)
print("RMSE:", rmse)

MAPE: 7.4288266575026505
MAE: 0.48619155696432614
MSE: 0.407708437669504
RMSE: 0.6385205068511927


MAPE of 7.42% indicates on average the absolute percentage difference between predicted and actual values is 7.42%.

MAE of .486 indicates on average the model predictions are off by about 0.486 units from the actual values.

MSE of .4077 represents average squared deviation of predictions from actual values.

RMSE of .63 indicates std of the residuals (prediction errors).

**2. Using Selected feature names from f_regression and SelectKBest**

Import libraries and read/display .csv

In [ ]:
import pandas as pd
import numpy as np

# Create a DataFrame
df = pd.read_csv('HBAT-200-Data.csv')

# Display the first few rows of the dataset
df.head()

ID  X1  X2  X3  X4  X5   X6   X7   X8   X9  ...  X14  X15  X16  X17  X18  \
0   1   2   0   1   1   1  8.5  3.9  2.5  5.9  ...  4.7  4.3  5.0  5.1  3.7   
1   2   3   1   0   0   0  8.2  2.7  5.1  7.2  ...  5.5  4.0  3.9  4.3  4.9   
2   3   3   0   1   1   1  9.2  3.4  5.6  5.6  ...  6.2  4.6  5.4  4.0  4.5   
3   4   1   1   1   1   0  6.4  3.3  7.0  3.7  ...  7.0  3.6  4.3  4.1  3.0   
4   5   2   0   1   0   1  9.0  3.4  5.2  4.6  ...  6.1  4.5  4.5  3.5  3.5   

   X19  X20  X21   X22  X23  
0  8.2  8.0  8.4  65.1    1  
1  5.7  6.5  7.5  67.1    0  
2  8.9  8.4  9.0  72.1    1  
3  4.8  6.0  7.2  40.1    0  
4  7.1  6.6  9.0  57.1    0  

[5 rows x 24 columns]

Find the target (y)

In [ ]:
# Calculate correlations between all columns
correlations = df.corr()

# Identify potential features based on correlation with other columns
# Exclude highly correlated columns and non-numeric columns like 'ID'
exclude_columns = ['ID']

# Filter out non-numeric columns and excluded columns
numeric_columns = df.select_dtypes(include=[np.number]).columns
filtered_features = [col for col in numeric_columns if col not in exclude_columns]

# Determine the target variable
# Find the column with the highest average correlation with other columns
avg_correlations = correlations.abs().mean()
target_variable = avg_correlations.idxmax()

# Define the feature matrix X and target variable y
X = df[filtered_features]
y = df[target_variable]

# Display the identified features and target variable
print("\nIdentified Features:")
print(filtered_features)
print("\nTarget Variable:")
print(target_variable)


Identified Features:
['X1', 'X2', 'X3', 'X4', 'X5', 'X6', 'X7', 'X8', 'X9', 'X10', 'X11', 'X12', 'X13', 'X14', 'X15', 'X16', 'X17', 'X18', 'X19', 'X20', 'X21', 'X22', 'X23']

Target Variable:
X19


Target (y) is X19.

What factors influence X19 Satisfaction?

Possibly:<br>
X6-Product Quality<br>
X7-E-Commerce<br>
X8-Technical Support<br>
X9-Complaint Resolution<br>
X10-Advertising<br>
X11-Product Line<br>
X12-Salesforce Image<br>
X13-Competitive Pricing<br>
X14-Warranty & Claims<br>
X15-New Products<br>
X16-Order & Billing<br>
X17-Price Flexibility<br>
X18-Delivery Speed<br>

Print variance

In [ ]:
columns_of_interest = ['X6', 'X7', 'X8', 'X9', 'X10', 'X11', 'X12', 'X13', 'X14', 'X15', 'X16', 'X17', 'X18']

# Calculate variance for each feature in columns_of_interest
variances = df[columns_of_interest].var()

# Print the variances
print("Variance for each feature:")
print(variances)


Variance for each feature:
X6     1.912728
X7     0.591231
X8     2.739649
X9     1.464014
X10    1.315746
X11    1.735553
X12    1.273664
X13    2.500361
X14    0.766229
X15    2.238030
X16    0.831594
X17    1.422416
X18    0.561552
dtype: float64


Correlation

In [ ]:
# Include 'X19' in columns of interest
columns_of_interest_with_target = columns_of_interest + ['X19']

# Select the columns of interest from your DataFrame
selected_df = df[columns_of_interest_with_target]

# Calculate the correlation matrix
correlation_matrix = selected_df.corr()

# Print the correlation of all features with the target 'X19'
target_correlation = correlation_matrix['X19'].sort_values(ascending=False)
print("Correlation of all features with 'X19':")
print(target_correlation)

Correlation of all features with 'X19':
X19    1.000000
X11    0.646337
X18    0.630172
X9     0.597566
X16    0.540485
X6     0.521052
X12    0.477870
X10    0.353541
X7     0.341551
X14    0.269225
X8     0.202572
X15    0.191509
X17    0.031824
X13   -0.282063
Name: X19, dtype: float64


Split the data into training and testing sets.

In [ ]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.feature_selection import SelectKBest, f_regression
from sklearn.metrics import mean_absolute_error, mean_absolute_percentage_error

# Prepare X and y
X = df[columns_of_interest].values  # Use columns_of_interest excluding 'X19'
y = df['X19'].values

# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.4, random_state=42)


Apply SelectKBest and train

In [ ]:
# Apply SelectKBest with f_regression to select the best features from the training set
selector = SelectKBest(score_func=f_regression, k=6)  # Use 'all' or adjust 'k' to select the number of features you want
X_train_selected = selector.fit_transform(X_train, y_train)
X_test_selected = selector.transform(X_test)

# Fit a linear regression model
model = LinearRegression()
model.fit(X_train_selected, y_train)

# Predict on the test set
y_pred = model.predict(X_test_selected)

y_pred

array([7.48297214, 7.35178717, 5.907771  , 5.99450322, 9.55167345,
       7.70710647, 6.27750051, 5.94418187, 6.44102601, 6.87746328,
       6.41221126, 9.61302734, 6.86264871, 8.80126721, 5.22658004,
       8.61172066, 8.89538854, 5.33272278, 6.06412925, 5.93020758,
       5.38373432, 6.33227339, 7.40342011, 8.8026677 , 7.25461625,
       7.33242228, 6.97357549, 5.97311826, 7.09529418, 4.31216174,
       6.9121608 , 6.61808029, 7.63498913, 7.6454034 , 8.51372391,
       7.36652124, 7.9715162 , 7.90017251, 6.93469105, 6.65948491,
       6.08571769, 8.7002772 , 7.46211354, 6.17316582, 5.05068917,
       7.09266119, 6.17144883, 8.87357929, 7.52595526, 4.66436143,
       7.31609935, 7.36438516, 5.07906672, 7.76239852, 6.40692395,
       8.69952648, 6.58363265, 7.86844519, 8.80497012, 6.37482701,
       6.87002449, 5.29045301, 5.7779413 , 9.00566323, 5.40421894,
       5.69660294, 8.65968031, 7.57575869, 7.28131643, 4.59430862,
       7.07776356, 6.42068687, 7.16427012, 7.75085689, 6.79375

Evaluate Model Performance - Mape/Mae/Mse/Rmse

In [ ]:
from sklearn.metrics import mean_absolute_error, mean_absolute_percentage_error, mean_squared_error

# MAPE
mape = (mean_absolute_percentage_error(y_test, y_pred) * 100)
print("MAPE:", mape)

#MAE
mae = mean_absolute_error(y_test, y_pred)

print("MAE:", mae)

#MSE (Mean Squared Error)
mse = mean_squared_error(y_test, y_pred)
print("MSE:", mse)


#RMSE (Root Mean squared error)
rmse = np.sqrt(mse)
print("RMSE:", rmse)

MAPE: 7.674749464438069
MAE: 0.49093719059028273
MSE: 0.41062306505979607
RMSE: 0.6407987711128947


MAPE of 7.67 indicates on average the absolute percentage difference between predicted and actual values is 7.67%.

MAE of .491 indicates on average the model predictions are off by about 0.491 units from the actual values.

MSE of .411 represents average squared deviation of predictions from actual values.

RMSE of .640 indicates std of the residuals (prediction errors).

View selected features

In [ ]:
# Get the indices of the selected features
selected_feature_indices = selector.get_support(indices=True)

# Get names of selected features from columns_of_interest
selected_feature_names = [columns_of_interest[i] for i in selected_feature_indices]

# Print selected feature names
print("Selected feature names:")
print(selected_feature_names)

Selected feature names:
['X6', 'X9', 'X11', 'X12', 'X16', 'X18']


**Observation**:

Between the two sets, based on metrics, the first set of features ['X6', 'X7', 'X11', 'X12', 'X13', 'X17'] may perform better in terms of predictive accuracy compared to the second attempt. The first set has lower MAPE, MAE, MSE, RMSE.